src/sqlfluff/core/parser/segments/base.py

```diff
@@ -1092,10 +1092,32 @@ def apply_fixes(self, dialect, rule_code, fixes):
 
     def _validate_segment_after_fixes(self, rule_code, dialect, fixes_applied, segment):
         """Checks correctness of new segment against match or parse grammar."""
-        found_error = False
         root_parse_context = RootParseContext(dialect=dialect)
         with root_parse_context as parse_context:
-            if getattr(segment, "match_grammar", None):
+            # Check for parse_grammar first. If both are present, it is more
+            # thorough and complete.
+            if getattr(segment, "parse_grammar", None):
+                try:
+                    # :HACK: Calling parse() corrupts the segment 'r'
+                    # in some cases, e.g. adding additional Dedent child
+                    # segments. Here, we work around this by calling
+                    # parse() on a "backup copy" of the segment.
+                    r_copy = deepcopy(segment)
+                    for seg in r_copy.segments:
+                        seg.pos_marker = replace(
+                            seg.pos_marker,
+                            templated_file=self.pos_marker.templated_file,
+                        )
+                    r_copy.parse(parse_context)
```

Comment:
This seems like the correct thing to do, even for `match_grammar`.

Why do we `parse` here, but `match` below? As per https://docs.sqlfluff.com/en/stable/architecture.html#stage-3-the-parser:

> Each segment can specify a parse_grammar, and a match_grammar. If both a match_grammar and parse_grammar are defined in a segment, match_grammar is used to quickly prune the tree for branches which do not match segments being parsed, and the parse_grammar is then used to refine the branch identified as correct. If only a match_grammar is defined, then it serves the purpose of both pruning and refining.

So even though it's ` match_grammar` shouldn't we still use it to `parse` in the case where there is no explicit `parse_grammar`? Surely a `match` check is only a partial check (and an easier one to pass than a full `parse`)?

Reply:
I'll try that. I wrote the original code here without much knowledge of the parser or input from those who do know it. 😬

Reply:
Seems to work. Also much simpler code. 👍